File: langgraph/src/graphs/how-it-works/nodes/process-next.ts
import type { State } from "../graph.js";

export async function processNextNode(state: State): Promise<Partial<State>> {
  const [next, ...rest] = state.pendingApps;

  if (!next) {
    // No more apps — signal END via currentApp: null
    return { currentApp: null, pendingApps: [] };
  }

  console.log(`\n${"━".repeat(58)}`);
  console.log(`  Processing: ${next.name}`);
  console.log(`${"━".repeat(58)}\n`);

  return {
    currentApp: next,
    pendingApps: rest,
    // Reset transient per-app state
    currentFiles: [],
    currentAnalysis: "",
    currentData: null,
  };
}

File: pypackages/how_it_works/src/how_it_works/nodes/process_next.py
"""Process-next node — pops the next app from the queue."""

from __future__ import annotations

from typing import Any


async def process_next_node(state: dict[str, Any]) -> dict[str, Any]:
    pending = state.get("pending_apps", [])

    if not pending:
        return {"current_app": None, "pending_apps": []}

    next_app, *rest = pending

    print(f"\n{'━' * 58}")
    print(f"  Processing: {next_app.name}")
    print(f"{'━' * 58}\n")

    return {
        "current_app": next_app,
        "pending_apps": rest,
        "current_files": [],
        "current_analysis": "",
        "current_data": None,
    }
